File: python/11566.py
import sys, math

sys.stdin = open("11566.txt")


while True:
	a = sys.stdin.readline().split()

	N,x,T,K = int(a[0]),int(a[1]),int(a[2]),int(a[3])

	# (cost + tea) * 1.1 = x
	# x / 1.1 - tea = cost
	max_cost = math.floor((x * (N+1)) / 1.1 - T * (N + 1))
	# print(max_cost)

	if N == 0:
		break

	items = []
	for i in range(K):
		a = sys.stdin.readline().split()
		items.append((int(a[0]), sum(list(map(int, a[1:])))))

	# dp[K][max_cost][N*2] 
	dp = [[[0 for i in range(2*(N+1)+1)] for j in range(max_cost+1)] for k in range(K+1)]
	# print(dp)
	for i in range(1, K+1):
		for j in range(max_cost+1):
			for k in range(2*(N+1)+1):
				if items[i-1][0] > j:
					dp[i][j][k] = dp[i-1][j][k]
				else:
					dp[i][j][k] = max(dp[i-1][j][k], dp[i-1][j-items[i-1][0]][k-1] + items[i-1][1])
	print(round(dp[K][max_cost][2*(N+1)] / (N+1), 2))

File: c++/11566.cc
#include <bits/stdc++.h>

using namespace std;

int main() 
{
	ifstream cin("11566.txt");

	while (true) {
		int N,x,T,K;
		cin >> N >> x >> T >> K;

		// cout << N << " " << x << " " << T << " " << K << endl;

		if (N == 0) {break;}

		int max_cost = floor((double)(x * (N + 1)) / 1.1 + 1e-6) - (N + 1) * T;

		vector<pair<int, int>> items;
		for (int i{0}; i<K; ++i) {
			int value, sum, temp;
			sum = 0;
			cin >> value;
			for (int j{0}; j<=N; ++j) {
				cin >> temp;
				sum += temp;
			}

			items.push_back(make_pair(value, sum));
			items.push_back(make_pair(value, sum));
		}



		vector<vector<vector<int>>>dp(250,vector<vector<int>>(1200,vector<int>(30,0)));
		for (int i{1}; i<=items.size(); ++i) {
			for (int j{0}; j<=max_cost; ++j) {
				for (int k{1}; k<=2*(N+1); ++k) {
					if (items[i-1].first > j) {
						dp[i][j][k] = dp[i-1][j][k];
					} else {
						dp[i][j][k] = max(dp[i-1][j][k], 
							dp[i-1][j-items[i-1].first][k-1] + items[i-1].second);
					}
				}
			}
		}
		cout << fixed << setprecision(2) << (double)(dp[items.size()][max_cost][2*(N+1)]) / (N+1) << endl;
	}

}